esp/public/media/scripts/program/modules/autoschedulerfrontend.js

```diff
@@ -0,0 +1,116 @@
+                        var stats_row = $j("<tr>");
+                        entry.forEach(function (el) {
+                            var stats_cell = $j("<td valign='top'>");
+                            label = el[0];
+                            lines = el[1];
+                            stats_cell.append('<b>' + label + '</b>');
+                            var bullets = $j('<ul>');
+                            lines.forEach(function(line) {
+                                bullets.append('<li>' + line + '</li>');
+                            });
+                            stats_cell.append(bullets);
+                            stats_row.append(stats_cell);
+                        });
+                        stats_table.append(stats_row);
+                    });
+                    stats_div.append(stats_table);
+                    stats_table_html = stats_table.html();
+                    stats_div.append("<b>Hit `Save room assignments' if you are happy with this. If not, you can try playing with the parameters, or locking some sections in the AJAX scheduler (but wait a couple seconds before re-running if you do), or you can just go schedule something yourself with the AJAX scheduler. Note that these assignments will be lost if you run the automatic scheduling assistant again.</b>");
+                }
+            }
+        },
+        error: autoschedulerErrorHandler,
+        dataType: 'json'
+    });
+}
+
+function saveAutoscheduler() {
+    $j('#autoschedulerinfo').html('Saving...');
+    var post_data = {'csrfmiddlewaretoken': csrf_token(), 'autoscheduler_data': autoscheduler_data};
+
+    $j.ajax({
+        url: "/manage/" + program_url_base + "/autoscheduler_save",
+        type: "post",
+        data: post_data,
+        success: function(data) {
+            data = data['response'][0];
+            stats_div = $j('#autoschedulerinfo');
+            if (data['error_msg'])
+            {
+                stats_div.html("A misconfiguration or unexpected situation prevented the autoscheduler from running: " + data['error_msg']);
+            } else {
+                stats_div.html("<h2>The following scheduling assignments have been saved successfully:</h2>");
+                stats_div.append(stats_table_html);
+            }
+        },
+        error: autoschedulerErrorHandler,
+        dataType: 'json'
+    });
+}
+
+function clearAutoscheduler() {
+    $j('#autoschedulerinfo').html('Clearing...');
+    var post_data = {'csrfmiddlewaretoken': csrf_token()};
+    autoscheduler_data = '';
+    stats_table_html = '';
+
+    $j.ajax({
+        url: "/manage/" + program_url_base + "/autoscheduler_clear",
+        type: "post",
+        data: post_data,
+        success: function() {
+            $j('#autoschedulerinfo').html("The scheduling assignments have been cleared successfully!");
+        },
+        error: autoschedulerErrorHandler,
+        dataType: 'json'
+    });
+}
+
+function toggleShowOptions(type) {
```

Comment:
Is jQuery's toggle https://api.jquery.com/toggle/ not enough?

I'm also not a fan of dynamically concatenating strings to make element IDs, it creates this action at a distance effect that makes the HTML source harder to read. If you just have `toggleShowOptions('autoschedulerconstraintsdiv')` in the HTML source I'm 99% confident I know exactly what it does without checking the JS.